src/diagnostics/MOM_harmonic_analysis.F90

```diff
@@ -362,74 +376,79 @@ subroutine HA_write(ha1, Time, G, CS)
 
 end subroutine HA_write
 
-!> This subroutine computes the harmonic constants (stored in FtSSHw) using the dot products of the temporal
+!> This subroutine computes the harmonic constants (stored in x) using the dot products of the temporal
 !! basis functions accumulated in FtF, and the dot products of the SSH (or other fields) with the temporal basis
 !! functions accumulated in FtSSH. The system is solved by Cholesky decomposition,
 !!
-!!     FtF * FtSSHw = FtSSH,    =>    FtFw * (FtFw' * FtSSHw) = FtSSH,
+!!     FtF * x = FtSSH,    =>    L * (L' * x) = FtSSH,    =>    L * y = FtSSH,
 !!
-!! where FtFw is a lower triangular matrix, and the prime denotes matrix transpose.
+!! where L is the lower triangular matrix, y = L' * x, and x is the solution vector.
 !!
-subroutine HA_solver(ha1, nc, FtF, FtSSHw)
+subroutine HA_solver(ha1, nc, FtF, x)
   type(HA_type), pointer,              intent(in)  :: ha1    !< Control structure for the current field
   integer,                             intent(in)  :: nc     !< Number of harmonic constituents
   real, dimension(:,:),                intent(in)  :: FtF    !< Accumulator of (F' * F) for all fields [nondim]
-  real, dimension(:,:,:), allocatable, intent(out) :: FtSSHw !< Work array for Cholesky decomposition [A]
+  real, dimension(:,:,:), allocatable, intent(out) :: x      !< Solution vector of harmonic constants [A]
 
   ! Local variables
-  real :: tmp0                              !< Temporary variable for Cholesky decomposition [nondim]
-  real, dimension(:),   allocatable :: tmp1 !< Temporary variable for Cholesky decomposition [nondim]
-  real, dimension(:,:), allocatable :: tmp2 !< Temporary variable for Cholesky decomposition [A]
-  real, dimension(:,:), allocatable :: FtFw !< Lower triangular matrix for Cholesky decomposition [nondim]
+  real :: tmp0                                !< Temporary variable for Cholesky decomposition [nondim]
+  real, dimension(:,:),   allocatable :: L    !< Lower triangular matrix of Cholesky decomposition [nondim]
+  real, dimension(:),     allocatable :: tmp1 !< Inverse of the diagonal entries of L [nondim]
+  real, dimension(:,:),   allocatable :: tmp2 !< 2D temporary array involving FtSSH [A]
+  real, dimension(:,:,:), allocatable :: y    !< 3D temporary array, i.e., L' * x [A]
   integer :: k, m, n, is, ie, js, je
 
   is = ha1%is ; ie = ha1%ie ; js = ha1%js ; je = ha1%je
 
+  allocate(L(1:2*nc+1,1:2*nc+1), source=0.0)
   allocate(tmp1(1:2*nc+1), source=0.0)
   allocate(tmp2(is:ie,js:je), source=0.0)
-  allocate(FtFw(1:2*nc+1,1:2*nc+1), source=0.0)
-  allocate(FtSSHw(is:ie,js:je,2*nc+1), source=0.0)
+  allocate(x(is:ie,js:je,2*nc+1), source=0.0)
+  allocate(y(is:ie,js:je,2*nc+1), source=0.0)
```

Comment:
@c2xu It was spotted by @Hallberg-NOAA that these could be declared as locals rather than allocated arrays.  I agree that it makes more sense to simply add the sizes in the definitions, e.g.
```
real :: L(2*nc+1, 2*nc+1)
real :: tmp1(2*nc+1)
real :: tmp2(is:ie, js:je)
real :: x(is:ie, js:je, 2*nc+1)
real :: y(is:ie, js:je, 2*nc+1)
```
Then initialize to zero if necessary (but not in the definition!).

It might be a bit problematic if `nc` is too large, but even then I'm probably just being overly cautious.

Reply:
Thanks. The code has been updated.